### PR TITLE
Config reload [WIP]

### PR DESCRIPTION
**Pull request by tylertreat**
This is a rough first pass at config reload (issue #338). This is just a tracer that implements reload support for the `trace` config option to prove out a working end-to-end example. If we like the overall approach, we can start implementing support for additional config options in a piecemeal fashion. This PR is mostly just to get early feedback.

**This PR does a few things:**
- Adds a `Reload` function to `Server` which can be called to reload configuration from a file.
- Adds support for reloading `trace` option. Attempting to change any other options will result in a reload failure.
- If reload fails, the server continues using the previous config.
- Adds a SIGHUP handler which invokes config reload.
- Some small refactors to reuse code and minimize changes.

**Open questions:**
- If the new config file is invalid, should we dump the last good config to a .bak file (or do something else)?
- How should we signal config reload on Windows?

@nats-io/core

### CODE REVIEW

**Comment by tylertreat** [comment]
Anyone know what's up with the unit tests? `go test` is returning with exit status 1 but it looks like all the tests passed.

**Comment by kozlovic** [comment]
All builds fail at "TestRouteTLSHandshakeError", you may want to look if your changes have introduced a call to os.Exit() or something that this test may trigger.

File: server/log.go [hunk 1]
```diff
@@ -15,9 +15,9 @@ var trace int32
 var debug int32
 
 var log = struct {
-	sync.Mutex
+	*sync.Mutex
```
Comment (derekcollison) [inline server/log.go]: Why this change?

File: server/server.go [hunk 2]
```diff
@@ -148,6 +147,18 @@ func New(opts *Options) *Server {
 	return s
 }
 
+func (s *Server) getOpts() *Options {
+	s.optsMu.RLock()
+	defer s.optsMu.RUnlock()
```
Comment (derekcollison) [inline server/server.go]: This might be expensive for when getOpts is in a fast path. Defer is better, but still slow. Just explicitly unlock like setOpts does.

File: server/signal.go [hunk 3]
```diff
@@ -11,12 +11,12 @@ import (
...
 // Signal Handling
 func (s *Server) handleSignals() {
-	if s.opts.NoSigs {
+	if s.getOpts().NoSigs {
 		return
 	}
 	c := make(chan os.Signal, 1)
 
-	signal.Notify(c, syscall.SIGINT, syscall.SIGUSR1)
+	signal.Notify(c, syscall.SIGINT, syscall.SIGUSR1, syscall.SIGHUP)
```
Comment (derekcollison) [inline server/signal.go]: I would not add this in until we have programmatic tests via API for all options we care about etc. and good test coverage. We should not expose to end users too soon IMO.

File: server/log.go (hunk 1 above)
Comment (tylertreat) [inline server/log.go]: Unit test is storing the value of `log` before the test and replacing it after to prevent side effects, so govet complains because it's copying the mutex value (motivation for #501).

File: server/reload.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+			oldValue = oldConfig.Field(i).Interface()
+			newValue = newConfig.Field(i).Interface()
+			changed  = !reflect.DeepEqual(oldValue, newValue)
+		)
+		if !changed {
+			continue
+		}
+		switch strings.ToLower(field.Name) {
+		case "trace":
+			opts = append(opts, &traceOption{newValue.(bool)})
```
Comment (petemiron) [inline server/reload.go]: I think it would be clearer to name `opts` something like `diff_opts`. took me a second to get why you were appending.

File: server/reload_test.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+		MaxControlLine: 1024,
+		MaxPayload:     1048576,
+		MaxConn:        65536,
+		PingInterval:   2 * time.Minute,
+		MaxPingsOut:    2,
+		WriteDeadline:  2 * time.Second,
+	}
+	processOptions(golden)
+
+	if err := os.Symlink("./configs/reload/test.conf", config); err != nil {
```
Comment (petemiron) [inline server/reload_test.go]: does this work as expected on windows? cc @ColinSullivan1

File: server/signal.go (hunk 3 above)
Comment (petemiron) [inline server/signal.go]: I agree with @derekcollison. let's build up the features for supporting it and then add in the reload once we have the critical items (Certs and Auth).

Derek, you agree we can merge all the bits to support in prior to having the signal and making the feature available in the server, correct?

File: server/reload_test.go (hunk 5 above)
Comment (tylertreat) [inline server/reload_test.go]: It should (https://golang.org/src/os/file_windows.go#L522), though I don't have a Windows box to test on.

Comment (ColinSullivan1) [inline server/reload_test.go]: It works on windows so long as you have correct privileges - you'll want to add a note someplace, probably the README.md.

Without the right privileges:
```--- FAIL: TestConfigReload (0.00s)
reload_test.go:175: Error creating symlink: symlink .\configs\reload\test.conf g:\src\github.com\nats-io\gnatsd\server\tmp.conf: A required privilege is not held by the client.
FAIL
```
On a network drive (generally not allowed by default):

```
--- FAIL: TestConfigReload (0.00s)
        reload_test.go:175: Error creating symlink: symlink .\configs\reload\test.conf g:\src\github.com\nats-io\gnatsd\server\tmp.conf: Incorrect function.
FAIL
```
Locally, with privileges (simple case running on `C:` as admin), you'll succeed.
```
=== RUN   TestConfigReload
[3792] [INF] Reloaded server configuration
--- PASS: TestConfigReload (0.00s)
```

**Comment by ColinSullivan1** [comment]
Regarding a config reload on windows. some other mechanism will be needed (that signal is not really supported).  Either a REST API, or something else.  We have the same issue with log rotation.

File: server/reload_test.go (hunk 5 above)
Comment (tylertreat) [inline server/reload_test.go]: @ColinSullivan1 Thanks for checking that. I can add a note to the README if we think this is acceptable. Since this is just for development purposes, this seems OK to me. What do you guys think?

**Comment by tylertreat** [comment]
@ColinSullivan1 Yep, since we're not exposing the signals yet, I think we can defer the decision for what we want to do on Windows for now. That can come at a later time when we're ready to expose the feature.

File: server/reload_test.go (hunk 5 above)
Comment (ColinSullivan1) [inline server/reload_test.go]: IMO this is OK; nothing unusual (at least for Windows).

File: server/reload_test.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+		t.Fatalf("Error creating symlink: %v", err)
```
Comment (ColinSullivan1) [inline server/reload_test.go]: I'm torn on this, but maybe skip the test here (and in others below), with some output?  This makes assumptions about the privileges of the user the test is running under, where the test is running, etc. It's convenience vs ensuring a test runs locally.  wdyt?

File: server/server.go [hunk 7]
```diff
@@ -584,7 +596,7 @@ func (s *Server) createClient(conn net.Conn) *client {
 	}
 	// If there is a max connections specified, check that adding
 	// this new client would not push us over the max
-	if s.opts.MaxConn > 0 && len(s.clients) >= s.opts.MaxConn {
+	if s.getOpts().MaxConn > 0 && len(s.clients) >= s.getOpts().MaxConn {
```
Comment (ColinSullivan1) [inline server/server.go]: Do you think it is worth optimizing in this function to snapshot MaxConn, TLS options, and AuthTimeout?  I realize establishing the connection and TLS are already expensive, but am thinking of a thundering herd of clients reconnecting.

File: server/reload_test.go (hunk 6 above)
Comment (tylertreat) [inline server/reload_test.go]: IMO it's better to fail loudly, otherwise the test might get skipped and the user wouldn't notice the warning message. We could add additional information to the Fatal message.

Comment (ColinSullivan1) [inline server/reload_test.go]: I can go with that...   IMO an "ensure you have privileges" would go a long way.

**Comment by derekcollison** [comment]
@tylertreat Let's rebase to master. Keep signalling removed and make sure we cover Windows builds, etc. Then we are good to go.

File: server/reload.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+// reloadOptions reloads the server config with the provided options. If an
+// option that doesn't support hot-swapping is changed, this returns an error.
+func (s *Server) reloadOptions(newOpts *Options) error {
+	changed, err := s.diffOptions(newOpts)
+	if err != nil {
+		return err
+	}
+	s.setOpts(newOpts)
+	s.applyOptions(changed)
+	return nil
```
Comment (tylertreat) [inline server/reload.go]: The problem with failing an option application is we could end up in an inconsistent state—some new options are applied, others are not—unless we also introduce a "revert" operation to revert any applied options.

Comment (kozlovic) [inline server/reload.go]: No because we apply the options to a temporary Options structure, and do not swap the server's options with the new one until after success.

Comment (tylertreat) [inline server/reload.go]: There are two parts to swapping `Options`: 1) swapping the `Options` pointer itself and 2) mutating the server as necessary (e.g. for `trace` we need to update the server logger. 2) could result in an inconsistent state of the server.

Comment (kozlovic) [inline server/reload.go]: I see, but if we apply the new options to the server's options that a caller has a pointer to (getOpts()), there is a risk that the caller accesses values that are being changed on the fly.

Comment (tylertreat) [inline server/reload.go]: Agreed, there is, but I'm not sure how we can apply a new config atomically.

**Comment by tylertreat** [comment]
I believe I've addressed everyone's feedback. @kozlovic can you take another look?

File: server/log.go [hunk 9]
```diff
@@ -58,12 +88,16 @@ func (s *Server) ReOpenLogFile() {
...
+
+	if opts.LogFile == "" {
 		s.Noticef("File log re-open ignored, not a file logger")
 	} else {
-		fileLog := logger.NewFileLogger(s.opts.LogFile,
-			s.opts.Logtime, s.opts.Debug, s.opts.Trace, true)
-		s.SetLogger(fileLog, s.opts.Debug, s.opts.Trace)
+		fileLog := logger.NewFileLogger(opts.LogFile,
+			opts.Logtime, s.getOpts().Debug, opts.Trace, true)
+		s.SetLogger(fileLog, s.getOpts().Debug, opts.Trace)
```
Comment (kozlovic) [inline server/log.go]: Still using `s.getOpts()` here...

**Comment by tylertreat** [comment]
@derekcollison I think this is good to go.
